### ISSUE TITLE
org: Add fontification, etc function

### ISSUE THREAD

**Issue by bdarcus** [emacs-citar/citar#641]
We might want to adapt `org-cite-basic-activate` to:

1. remove a dependency (it might be a little awkward here having that load the same bibliography file(s), etc.)
2. add some functionality (like previewing citation output when hovering over "cite")

We could also, or otherwise, send in a patch for `oc-basic` so it can do 2.

**Comment by bdarcus** [comment]
Now that you've merged citar cache support, @andras-simonyi, any thoughts on this?

**Comment by andras-simonyi** [comment]
Do I understand correctly that currently Citar's activation processor is the built-in org-cite "basic" plus a local keymap for citations?

**Comment by bdarcus** [comment]
Yes.

**Comment by andras-simonyi** [comment]
I'm thinking of adding this keymap to org-cite-csl-activate...

**Comment by bdarcus** [comment]
Cool.

If you do, any thoughts on what that might mean for the questions I raise in the OP?

Clearly one option is simply to remove this activate processor entirely, and recommend `org-cite-csl-activate`.

cc @roshanshariff @aikrahguzar @localauthor (no rush though, but I'm not sure if you've seen this issue)

**Comment by andras-simonyi** [comment]
I still think org-cite-csl-activate is a bit too experimental/idiosyncratic to be a good default. OTOH providing nice previews of cites and bibliography entries on hover is definitely doable reliably. Also, for a while I've been thinking about activating citations by selectively hiding/prettifying some parts: the cite part could be hidden or replaced by a symbol, and typical keys contain an author part and a year part which could be hidden depending on the citation style. Unfortunately, I didn't have a prototype yet.

**Comment by bdarcus** [comment]
> OTOH, providing nice previews of cites and bibliography entries on hover is definitely doable reliably.

Where do you think that belongs?

In `oc-basic-activate`? In `org-cite-csl-activate`?

Or perhaps `oc.el` might need some support for this too?

**Comment by andras-simonyi** [comment]
> > OTOH, providing nice previews of cites and bibliography entries on hover is definitely doable reliably.
> 
> Where do you think that belongs?
> 
> In `oc-basic-activate`? In `org-cite-csl-activate`?

Good question, oc-basic probably shouldn't depend on citeproc-el, so this leaves the second alternative, or a separate module. OTOH, within the context of citar, it'd be logical to use the citar cache, so could also remain a citar module too. Code-wise I'd expect it to be rather short, a few dozen LOC.

**Comment by bdarcus** [comment]
> OTOH, within the context of citar, it'd be logical to use the citar cache, so could also remain a citar module too.

Right, but your activate processor already supports that, and I guess only part of the hover preview.

**Comment by andras-simonyi** [comment]
> Right, but your activate processor already supports that, and I guess only part of the hover preview (not yet the citation).

Well, I was thinking of factoring this functionality out into a module somehow, but another alternative would be indeed to say that the main functionality of csl-activate is the csl-based citation (and bibliography) preview and offering a choice between providing the preview as a text-replacing display property (the current state of affairs) or as a tooltip over a basic-like activation.

**Comment by bdarcus** [comment]
@andras-simonyi 

> Well, I was thinking of factoring this functionality out into a module somehow, 

A module for what? `oc-csl`?

> but another alternative would be indeed to say that the main functionality of csl-activate is the csl-based citation (and bibliography) preview and offering a choice between providing the preview as a text-replacing display property (the current state of affairs) or as a tooltip over a basic-like activation.

Right; that's what I was thinking.

You can have the default be tooltips while you still consider the current behavior still experimental?

**Comment by lazzalazza** [comment]
As for the previewing part, I have written this code that uses eldoc to output a simple preview of the content of the citation.
It works pretty smoothly: don't take into account the coloring part, but it might be easily implementable. 

(defun citar-eldoc-entry (callback &rest _ignored)
  "Return the definition of the citation at point for Eldoc."
  (when-let (citation (citar-citation-at-point))
    (let* ((citations  (car citation))
           (entry (citar-get-entry citations))
           (author (or (citar-get-value "author" entry)
                       (citar-get-value "editor" entry)))
           (title (citar-get-value "title" entry))
           (date (or (citar-get-value "date" entry)
                     (citar-get-value "year" entry))))
        (funcall callback
                 (format "%s" (concat (if author
                                          (propertize (concat author ", ") 'face '(:foreground "red")))
                                      (propertize title 'face '(:foreground "green"))
                                      (propertize (concat ", " date) 'face '(:foreground "yellow"))))
                 ;; :thing 'footnote-definition
                 :face 'font-lock-doc-face)
      citation)))

  (defun citar-enable-eldoc ()
    "Enable Eldoc support for citations."
    (add-hook 'eldoc-documentation-functions #'citar-eldoc-entry nil t))

  (add-hook 'org-mode-hook #'citar-enable-eldoc)

**Comment by lazzalazza** [comment]
(It would be great to also use `citar-citeproc-format-reference` to output a preview, but I am apparently getting errors from the XML parser...: "xml-parse-attlist: XML: (Not Well-Formed) Attribute values must be given between quotes")

**Comment by dschaehi** [comment]
> As for the previewing part, I have written this code that uses eldoc to output a simple preview of the content of the citation. It works pretty smoothly: don't take into account the coloring part, but it might be easily implementable.
> 
> ```
> (defun citar-eldoc-entry (callback &rest _ignored)
>   "Return the definition of the citation at point for Eldoc."
>   (when-let (citation (citar-citation-at-point))
>     (let* ((citations  (car citation))
>            (entry (citar-get-entry citations))
>            (author (or (citar-get-value "author" entry)
>                        (citar-get-value "editor" entry)))
>            (title (citar-get-value "title" entry))
>            (date (or (citar-get-value "date" entry)
>                      (citar-get-value "year" entry))))
>         (funcall callback
>                  (format "%s" (concat (if author
>                                           (propertize (concat author ", ") 'face '(:foreground "red")))
>                                       (propertize title 'face '(:foreground "green"))
>                                       (propertize (concat ", " date) 'face '(:foreground "yellow"))))
>                  ;; :thing 'footnote-definition
>                  :face 'font-lock-doc-face)
>       citation)))
> 
>   (defun citar-enable-eldoc ()
>     "Enable Eldoc support for citations."
>     (add-hook 'eldoc-documentation-functions #'citar-eldoc-entry nil t))
> 
>   (add-hook 'org-mode-hook #'citar-enable-eldoc)
> ```

Thanks for the snippet! This, however, does not seem to work when there are multiple citations (e.g., `[cite:@chen_what_2024;@ye_physics_2024;@saunshi_inductive_2024]`). Even if I move the cursor to the second citation, only the preview of the first citation is shown in the minibuffer. Can you confirm this behavior?
